src/rules/no-element-overwrite.ts

```diff
@@ -0,0 +1,162 @@
+import { Rule, AST } from "eslint";
+import {
+  Node,
+  SwitchCase,
+  BlockStatement,
+  Statement,
+  AssignmentExpression,
+  MemberExpression,
+  Expression,
+  Program,
+  ModuleDeclaration,
+} from "estree";
+import { areEquivalent } from "../utils/equivalence";
+import {
+  isExpressionStatement,
+  isMemberExpression,
+  isAssignmentExpression,
+  isLiteral,
+  isIdentifier,
+  isCallExpression,
+} from "../utils/nodes";
+
+const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    return {
+      SwitchCase(node: Node) {
+        const switchCase = node as SwitchCase;
+        checkStatements(switchCase.consequent);
+      },
+
+      BlockStatement(node: Node) {
+        const block = node as BlockStatement;
+        checkStatements(block.body);
+      },
+
+      Program(node: Node) {
+        const program = node as Program;
+        checkStatements(program.body);
+      },
+    };
+
+    function checkStatements(statements: Array<Statement | ModuleDeclaration>) {
+      const usedKeys: Map<string, KeyWriteCollectionUsage> = new Map();
+      let collection: Node | null = null;
+      statements.forEach(statement => {
+        const keyWriteUsage = getKeyWriteUsage(statement);
+        if (keyWriteUsage) {
+          if (collection && !areEquivalent(keyWriteUsage.collectionNode, collection, context.getSourceCode())) {
+            usedKeys.clear();
+          }
+          const sameKeyWriteUsage = usedKeys.get(keyWriteUsage.indexOrKey);
+          if (sameKeyWriteUsage) {
+            context.report({
+              node: keyWriteUsage.node,
+              message: 'Verify this is the index that was intended; "{{index}}" was already set on line {{line}}.',
+              data: {
+                index: keyWriteUsage.indexOrKey,
+                line: String(sameKeyWriteUsage.node.loc!.start.line),
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Forbidden non null assertion ([typescript:S2966](https://sonarcloud.io/organizations/sonarsource/rules#rule_key=typescript%3AS2966))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=eslint-plugin-sonarjs&pullRequest=36&issues=AWK0wBF4jr7wspRoK9BA&open=AWK0wBF4jr7wspRoK9BA)

[ISSUE_KEY]: AWK0wBF4jr7wspRoK9BA